### ISSUE TITLE
Erase elements of propositions from the memory

### ISSUE THREAD

**Issue by guillaumebrunerie** [agda/agda#3510]
I’m experiencing very high memory usage (and compilation time) in my current project. I’m not really sure why but I’m suspecting it might come from the large proofs (in Prop) that are kept in memory.

For instance when I normalize one of the goals with implicit arguments on, the result takes more than 3 million characters, but when I then remove by hand all the subterms that are actually elements of a proposition, then it only takes 700 characters.

Would it make sense to say that any subterm whose type has sort Prop should be simply removed, to save memory? (i.e. replaced by some dummy term, or something like that I guess)
It would also probably make it easy (or even mandatory) to fix #3337.

**Comment by jespercockx** [comment]
If we do this, we would still have to keep track of the free variables that occurred in the erased term. Possibly there is also other pieces of information that we should remember. @andreasabel I seem to remember that irrelevant arguments were being erased at some point but that patch was eventually reverted, do you remember the reason for that?

**Comment by andreasabel** [comment]
Further reading e.g. #337 and #483. Irrelevant terms were retained because they were needed for:
- `with`-abstraction (as it is untyped, it may become unsound with erasure)
- termination checking
- meta-variable solving (as @jespercockx says, we would need to retain at least the free variables)

**Comment by guillaumebrunerie** [comment]
Now that I’m thinking about it, I think the issue is rather that in some cases Agda will attempt to normalize elements of propositions, which should never be needed.
I’ll see if I can figure out in which cases it does that.

**Comment by jespercockx** [comment]
Probably this happens when you're using `--rewriting` since the rewrite matcher is currently not type-aware. I'm actually working on changing that at the moment, but it requires some rather big refactorings.

**Comment by guillaumebrunerie** [comment]
Ah, this is indeed happening in a place where rewriting is used a lot. Nice to hear that you’re working on it, I’ll try to find a workaround in the mean time.

**Comment by gallais** [comment]
If the issue is coming from proofs of equality, you could erase them yourself by using
`primEraseEquality`. Even though we ship the usual version of equality & erasure
the primitive is compatible with a `Prop`-based equality too:

```agda
{-# OPTIONS --prop #-}

module eraseProp where

data _≡_ {a} {A : Set a} (a : A) : A → Prop where
  refl : a ≡ a

{-# BUILTIN EQUALITY _≡_ #-}

primitive

  primEraseEquality : ∀ {a} {A : Set a} {x y : A} → x ≡ y → x ≡ y


-- It does indeed erase to refl when x is syntactically equal to y:

data _≈_ {P : Prop} (p : P) : P → Prop where
  refl : p ≈ p

_ : ∀ {a} {A : Set a} {x : A} (p : x ≡ x) → primEraseEquality p ≈ refl
_ = λ p → refl
```

**Comment by guillaumebrunerie** [comment]
I’m not sure if that would do what I want, most of the time `x` is not syntactically equal to `y`.
What happens if you ask Agda to normalize `primEraseEquality p` in that case? Won’t it go inside and attempt to normalize `p`?

One way that does seem to work is as follows. The postulate is a unicode symbol that you’re not supposed to use, and `# p` is the same thing as `p` except that it kills `p` whenever Agda tries to reduce it. So the idea is simply to use `#` like some kind of `erase`.
```agda

postulate
  ‗ : {A : Prop} → A

# : {A : Prop} → A → A
# a = ‗
```
An issue with that, though, is that it is trivial to prove false by exploiting the termination checker…
```agda
oops : False
oops = # oops
```

**Comment by gallais** [comment]
> What happens if you ask Agda to normalize `primEraseEquality p` in that case? Won’t it go inside and attempt to normalize `p`?

It won't ([primitive](https://github.com/agda/agda/blob/1becd2e22feae7a0a5635b7431e68089d3ba5d98/src/full/Agda/TypeChecking/Primitive.hs#L1630), [abstract machine](https://github.com/agda/agda/blob/master/src/full/Agda/TypeChecking/Reduce/Fast.hs#L621)). If it did it would be a bug.

You can experiment by adding this buggy program at the end of my previous example:

```agda
module _ {a} {A : Set a} (x : A) where

  {-# TERMINATING #-}
  argh : ∀ y → x ≡ y
  argh y = primEraseEquality (argh y)

  _ : argh x ≈ refl
  _ = refl

open import Agda.Builtin.Nat

p : 2 ≡ 3
p = argh 2 3
```

Normalising `p` yields `.(argh 2 3)`.

**Comment by guillaumebrunerie** [comment]
Ah ok, cool!
But in any case, I’m not sure it’s equalities that get reduced anyway, I have another datatype in Prop that I suspect is the culprit, but it’s not easy to figure out what exactly gets reduced but shouldn’t.

**Comment by guillaumebrunerie** [comment]
There are parts of my development which do not use any rewriting rules, but still use way less memory when I use the `#` hack described above, so I suspect there are other cases where elements of propositions are reduced/normalized.
I’m working on a small test case…

**Comment by guillaumebrunerie** [comment]
Ok, here is an instance of weird normalization:

```agda
{-# OPTIONS --prop #-}

open import Agda.Builtin.Equality

_∙_ : {X : Set} {a b c : X} → a ≡ b → b ≡ c → a ≡ c
_∙_ refl refl = refl

postulate
  magic : {X : Prop} → X

# : {X : Prop} → X → X
# x = magic

postulate
  A : Set
  #SHOULDSTAYHIDDEN# : A

F : A → A
F a = #SHOULDSTAYHIDDEN#

postulate
  P : Prop
  f : P → A
  a₀ : A
  g : (p : P) → F (f p) ≡ a₀

test : P → a₀ ≡ a₀
test p = {!!} ∙ g (# p)
```

The type of the hole is displayed as `?0 : a₀ ≡ #SHOULDSTAYHIDDEN#`, i.e. the left-hand side of `g (# p)`, which is supposed to be `F (f (# p))` somehow got normalized.
This does not happen with either `g p` or `g magic`.

**Comment by gallais** [comment]
Note that if you add `{-# NOINLINE # #-}` below the definition of `#` then you get
the goal that you wanted. We had a similar problem in #3486 with automatic inlining
which is important to have for efficiency but is sometimes not what you want.

**Comment by guillaumebrunerie** [comment]
Indeed, but I’m a bit confused. If `#` is automatically inlined, why don’t I get simply `F (f magic)` instead?

**Comment by guillaumebrunerie** [comment]
This seems related to Prop, if we change all the `Prop` to `Set`, then we do get `F (f magic)` (or `F (f (# p))` with `{-# NOINLINE # #-}`).

**Comment by guillaumebrunerie** [comment]
Here is a slightly different issue, but related:
```agda
{-# OPTIONS --prop #-}

open import Agda.Builtin.Equality

_∙_ : {X : Set} {a b c : X} → a ≡ b → b ≡ c → a ≡ c
_∙_ refl refl = refl

postulate
  magic : {X : Prop} → X

# : {X : Prop} → X → X
# x = magic

postulate
  A : Set
  #SHOULDSTAYHIDDEN# : A → A

F : A → A
F a = #SHOULDSTAYHIDDEN# a

{-# NOINLINE F #-}

postulate
  P : Prop
  f : P → A
  a₀ : A
  g : (p : P) → F (f p) ≡ a₀

test : P → a₀ ≡ a₀
test p =  {!!} ∙ g (# p)
```
The difference with the previous code is that now `#SHOULDSTAYHIDDEN#` takes an argument.
This now fails with the error
```
Cannot instantiate the metavariable _31 to solution
#SHOULDSTAYHIDDEN# (f .(magic)) since (part of) the solution was
created in an irrelevant context.
when checking that the expression g (# p) has type _b_31 ≡ a₀
```
(which is an error I see regularly, by the way)

It works fine with `{-# NOINLINE # #-}`.

So my guess is that in the first example, automatic inlining will inline `# p` to `.(magic)` (whatever that means) instead of inlining it to `magic`, but Agda doesn’t like it so it will normalize `F` in order to get rid of that `.(magic)`. In the second example here, normalizing `F` is not enough, so we get an (unexpected) error.

**Comment by guillaumebrunerie** [comment]
Is there an easy way to disable all inlining, or at least figure out which functions get inlined? I couldn't find any debug output related to inlining in the source code.

**Comment by nad** [comment]
Automatic inlining can be turned off using `--no-auto-inline`.

**Comment by guillaumebrunerie** [comment]
@nad Thanks, I somehow missed that.
